<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>PyRanges documentation</title>
  <meta name="description" content="PyRanges documentation">
  <meta name="generator" content="bookdown 0.5 and GitBook 2.6.7">

  <meta property="og:title" content="PyRanges documentation" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="PyRanges documentation" />
  
  
  

<meta name="author" content="Endre Bakken Stovner">


<meta name="date" content="2019-07-24">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="methods-for-manipulating-single-pyranges.html">
<link rel="next" href="intersecting-ranges.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction to PyRanges</a></li>
<li class="chapter" data-level="2" data-path="loadingcreating-pyranges.html"><a href="loadingcreating-pyranges.html"><i class="fa fa-check"></i><b>2</b> Loading/Creating PyRanges</a></li>
<li class="chapter" data-level="3" data-path="writing-pyranges-to-disk.html"><a href="writing-pyranges-to-disk.html"><i class="fa fa-check"></i><b>3</b> Writing PyRanges to disk</a></li>
<li class="chapter" data-level="4" data-path="subsetting-pyranges.html"><a href="subsetting-pyranges.html"><i class="fa fa-check"></i><b>4</b> Subsetting PyRanges</a></li>
<li class="chapter" data-level="5" data-path="manipulating-the-data-in-pyranges.html"><a href="manipulating-the-data-in-pyranges.html"><i class="fa fa-check"></i><b>5</b> Manipulating the data in PyRanges</a></li>
<li class="chapter" data-level="6" data-path="inserting-columns-based-on-overlap.html"><a href="inserting-columns-based-on-overlap.html"><i class="fa fa-check"></i><b>6</b> Inserting columns based on overlap</a></li>
<li class="chapter" data-level="7" data-path="concatenating-pyranges.html"><a href="concatenating-pyranges.html"><i class="fa fa-check"></i><b>7</b> Concatenating PyRanges</a></li>
<li class="chapter" data-level="8" data-path="piping-pyranges.html"><a href="piping-pyranges.html"><i class="fa fa-check"></i><b>8</b> Piping PyRanges</a></li>
<li class="chapter" data-level="9" data-path="printing-pyranges.html"><a href="printing-pyranges.html"><i class="fa fa-check"></i><b>9</b> Printing PyRanges</a></li>
<li class="chapter" data-level="10" data-path="iterating-over-pyranges.html"><a href="iterating-over-pyranges.html"><i class="fa fa-check"></i><b>10</b> Iterating over PyRanges</a></li>
<li class="chapter" data-level="11" data-path="sorting-pyranges.html"><a href="sorting-pyranges.html"><i class="fa fa-check"></i><b>11</b> Sorting PyRanges</a></li>
<li class="chapter" data-level="12" data-path="summarizing-pyranges.html"><a href="summarizing-pyranges.html"><i class="fa fa-check"></i><b>12</b> Summarizing PyRanges</a></li>
<li class="chapter" data-level="13" data-path="methods-for-manipulating-single-pyranges.html"><a href="methods-for-manipulating-single-pyranges.html"><i class="fa fa-check"></i><b>13</b> Methods for manipulating single PyRanges</a></li>
<li class="chapter" data-level="14" data-path="applying-custom-methods-to-pairs-of-pyranges.html"><a href="applying-custom-methods-to-pairs-of-pyranges.html"><i class="fa fa-check"></i><b>14</b> Applying custom methods to pairs of PyRanges</a></li>
<li class="chapter" data-level="15" data-path="intersecting-ranges.html"><a href="intersecting-ranges.html"><i class="fa fa-check"></i><b>15</b> Intersecting Ranges</a></li>
<li class="chapter" data-level="16" data-path="overlapping-ranges.html"><a href="overlapping-ranges.html"><i class="fa fa-check"></i><b>16</b> Overlapping Ranges</a></li>
<li class="chapter" data-level="17" data-path="joining-ranges.html"><a href="joining-ranges.html"><i class="fa fa-check"></i><b>17</b> Joining Ranges</a></li>
<li class="chapter" data-level="18" data-path="finding-the-closest-intervals.html"><a href="finding-the-closest-intervals.html"><i class="fa fa-check"></i><b>18</b> Finding the closest intervals</a></li>
<li class="chapter" data-level="19" data-path="statistics-similarity-between-pyranges.html"><a href="statistics-similarity-between-pyranges.html"><i class="fa fa-check"></i><b>19</b> Statistics: similarity between PyRanges</a></li>
<li class="chapter" data-level="20" data-path="statistics-spatial-correlations-between-pyranges.html"><a href="statistics-spatial-correlations-between-pyranges.html"><i class="fa fa-check"></i><b>20</b> Statistics: spatial correlations between PyRanges</a></li>
<li class="chapter" data-level="21" data-path="turning-ranges-into-rles.html"><a href="turning-ranges-into-rles.html"><i class="fa fa-check"></i><b>21</b> Turning Ranges into RLEs</a></li>
<li class="chapter" data-level="22" data-path="an-introduction-to-rles.html"><a href="an-introduction-to-rles.html"><i class="fa fa-check"></i><b>22</b> An introduction to RLEs</a></li>
<li class="chapter" data-level="23" data-path="run-length-encoding-dicts.html"><a href="run-length-encoding-dicts.html"><i class="fa fa-check"></i><b>23</b> Run Length Encoding dicts</a></li>
<li class="chapter" data-level="24" data-path="subsetting-rles.html"><a href="subsetting-rles.html"><i class="fa fa-check"></i><b>24</b> Subsetting Rles</a></li>
<li class="chapter" data-level="25" data-path="subsetting-pyrles.html"><a href="subsetting-pyrles.html"><i class="fa fa-check"></i><b>25</b> Subsetting PyRles</a></li>
<li class="chapter" data-level="26" data-path="using-multiple-cores.html"><a href="using-multiple-cores.html"><i class="fa fa-check"></i><b>26</b> Using multiple cores</a></li>
<li class="chapter" data-level="27" data-path="genomicfeatures-methods-using-genomic-context.html"><a href="genomicfeatures-methods-using-genomic-context.html"><i class="fa fa-check"></i><b>27</b> GenomicFeatures: methods using genomic context</a></li>
<li class="chapter" data-level="28" data-path="fetching-gene-tracks-as-pyranges.html"><a href="fetching-gene-tracks-as-pyranges.html"><i class="fa fa-check"></i><b>28</b> Fetching gene tracks as PyRanges</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">PyRanges documentation</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="applying-custom-methods-to-pairs-of-pyranges" class="section level1">
<h1><span class="header-section-number">14</span> Applying custom methods to pairs of PyRanges</h1>
<p>By using the <code>apply</code>, <code>apply_pair</code> or <code>apply_chunks</code> methods, you can run custom methods on the dataframes in your PyRanges. The <code>apply</code> and <code>apply_chunks</code> methods takes a single dataframe, while <code>apply_pair</code> takes a pair of dataframes.</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="im">import</span> pyranges <span class="im">as</span> pr
chipseq <span class="op">=</span> pr.data.chipseq()
chipseq_background <span class="op">=</span> pr.data.chipseq_background()
<span class="kw">def</span> print_strands(df1, df2):
    <span class="bu">print</span>(df1.Chromosome.iloc[<span class="dv">0</span>], df1.Strand.iloc[<span class="dv">0</span>], df2.Strand.iloc[<span class="dv">0</span>])
    <span class="cf">return</span> df1.head(<span class="dv">5</span>)
result <span class="op">=</span> chipseq.apply_pair(chipseq_background, print_strands, strandedness<span class="op">=</span><span class="st">&quot;opposite&quot;</span>)</code></pre></div>
<pre><code>## chr1 + -
## chr1 - +
## chr2 + -
## chr2 - +
## chr3 + -
## chr3 - +
## chr4 + -
## chr4 - +
## chr5 + -
## chr5 - +
## chr6 + -
## chr6 - +
## chr7 + -
## chr7 - +
## chr8 + -
## chr8 - +
## chr9 + -
## chr9 - +
## chr10 + -
## chr10 - +
## chr11 + -
## chr11 - +
## chr12 + -
## chr12 - +
## chr13 + -
## chr13 - +
## chr14 + -
## chr14 - +
## chr15 + -
## chr15 - +
## chr16 + -
## chr16 - +
## chr17 + -
## chr17 - +
## chr18 + -
## chr18 - +
## chr19 + -
## chr19 - +
## chr20 + -
## chr20 - +
## chr21 + -
## chr21 - +
## chr22 + -
## chr22 - +
## chrX + -
## chrX - +
## chrY + -
## chrY - +</code></pre>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="bu">print</span>(result)</code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   | Start     | End       | Name       | Score     | Strand       |
## | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         | 212609534 | 212609559 | U0         | 0         | +            |
## | chr1         | 169887529 | 169887554 | U0         | 0         | +            |
## | chr1         | 216711011 | 216711036 | U0         | 0         | +            |
## | chr1         | 144227079 | 144227104 | U0         | 0         | +            |
## | ...          | ...       | ...       | ...        | ...       | ...          |
## | chrY         | 21751211  | 21751236  | U0         | 0         | -            |
## | chrY         | 7701983   | 7702008   | U0         | 0         | -            |
## | chrY         | 21910706  | 21910731  | U0         | 0         | -            |
## | chrY         | 22054002  | 22054027  | U0         | 0         | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 240 rows and 6 columns from 24 chromosomes.</code></pre>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="kw">def</span> set_start_to_zero(df):
    df.Start <span class="op">=</span> <span class="dv">0</span>
    <span class="cf">return</span> df
<span class="bu">print</span>(chipseq.<span class="bu">apply</span>(set_start_to_zero))</code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   | Start     | End       | Name       | Score     | Strand       |
## | (category)   | (int64)   | (int32)   | (object)   | (int64)   | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | chr1         | 0         | 212609559 | U0         | 0         | +            |
## | chr1         | 0         | 169887554 | U0         | 0         | +            |
## | chr1         | 0         | 216711036 | U0         | 0         | +            |
## | chr1         | 0         | 144227104 | U0         | 0         | +            |
## | ...          | ...       | ...       | ...        | ...       | ...          |
## | chrY         | 0         | 15224260  | U0         | 0         | -            |
## | chrY         | 0         | 13517917  | U0         | 0         | -            |
## | chrY         | 0         | 8010976   | U0         | 0         | -            |
## | chrY         | 0         | 7405401   | U0         | 0         | -            |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 10,000 rows and 6 columns from 24 chromosomes.</code></pre>
<p>Keyword arguments can be sent to the function:</p>
<div class="sourceCode"><pre class="sourceCode python"><code class="sourceCode python"><span class="kw">def</span> _print(df, kwargs):
    <span class="bu">print</span>(<span class="st">&quot;My keyword arguments were:&quot;</span>, kwargs.get(<span class="st">&quot;value&quot;</span>), <span class="st">&quot;and&quot;</span>, kwargs.get(<span class="st">&quot;whatever&quot;</span>))
    <span class="cf">return</span> df
chipseq.<span class="bu">apply</span>(_print, value<span class="op">=</span><span class="dv">123</span>, whatever<span class="op">=</span><span class="st">&quot;hi there!&quot;</span>)</code></pre></div>
<pre><code>## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!
## My keyword arguments were: 123 and hi there!</code></pre>
<p>They were printed 24 times since the function was applied once per chromosome.</p>
<p>apply_chunks splits each chromosome into <code>nb_cpu</code> chunks and runs each chunk in parallel, which is useful for slow row-based operations (like fisher exact, for example).</p>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="methods-for-manipulating-single-pyranges.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="intersecting-ranges.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
