# Manipulating the data in PyRanges

PyRanges is a thin wrapper around genomic data contained in pandas dataframes.
This dataframe is accessible with the df attribute of the PyRanges object.

```{python tidy=FALSE}
import pyranges as pr
gr = pr.data.chipseq()
print(gr)
print(gr.df.head(5))
```

To access a column of this dataframe, you can ask for the name directly from the
PyRanges object.

```{python tidy=FALSE}
print(gr.Start.head())
```

You can directly insert a column by setting the attribute on the PyRanges object:

```{python tidy=FALSE}
gr.stupid_example = "Hi There!"
print(gr)
gr = gr.drop("stupid_example")
print(gr)
```

And as you can see, you can drop columns from the PyRanges using drop. Using
drop by itself removes all metadata-columns. If you also want to remove the
strand you need to use the drop_strand flag.

All columns except Chromosome, Start, End and Strand can be changed in any way
you please and more metadata-columns can be added by setting it on the PyRanges
object. If you wish to change the Chromosome, Start, End and Strand columns you
should make a copy of the data from the PyRanges object and use it to
instantiate a new PyRanges object.

```{python tidy=FALSE}
import pandas as pd
gr.Name = gr.Chromosome.astype(str) + "_" + pd.Series(range(len(gr)), index=gr.Chromosome.index).astype(str)
print(gr)
```
