# Writing PyRanges to disk

The PyRanges can be written to several formats, namely csv, gtf and bigwig.

If no path-argument is given, the string representation of the data is returned.
(It may potentially be very large.) If a path is given, the return value is the
object itself. This way the write methods can easily be inserted in method call
chains.

```{python tidy=FALSE}
import pyranges as pr
import pyranges_db as pr_db
gr = pr.data.chipseq()

gr.to_gtf("chipseq.gtf")

print(gr[:10000].to_gtf())
```

The to_csv method takes the arguments header and sep.

```{python tidy=FALSE}
gr.to_csv("chipseq.csv", sep="\t", header=True)

print(gr[:10000].to_csv(sep="|", header=True))
```

The bigwig writer needs to know the chromosome sizes. You can fetch these using
the pyranges database functions.

```{python tidy=FALSE}
chromsizes = pr_db.ucsc.chromosome_sizes("hg19")
gr.to_bigwig("chipseq.bw", chromsizes)
```

If you want to write one bigwig for each strand, you need to do it manually.

```{python tidy=FALSE}
gr["+"].to_bigwig("chipseq_plus.bw", chromsizes)
gr["-"].to_bigwig("chipseq_minus.bw", chromsizes)
```
